pandas/_libs/interval.pyx

```diff
@@ -0,0 +1,194 @@
+cimport numpy as np
+import numpy as np
+import pandas as pd
+
+cimport util
+cimport cython
+import cython
+from numpy cimport *
+
+from cpython.object cimport (Py_EQ, Py_NE, Py_GT, Py_LT, Py_GE, Py_LE,
+                             PyObject_RichCompare)
+
+import numbers
+_VALID_CLOSED = frozenset(['left', 'right', 'both', 'neither'])
+
+cdef class IntervalMixin:
+    property closed_left:
+        def __get__(self):
+            return self.closed == 'left' or self.closed == 'both'
+
+    property closed_right:
+        def __get__(self):
+            return self.closed == 'right' or self.closed == 'both'
+
+    property open_left:
+        def __get__(self):
+            return not self.closed_left
+
+    property open_right:
+        def __get__(self):
+            return not self.closed_right
+
+    property mid:
+        def __get__(self):
+            try:
+                return 0.5 * (self.left + self.right)
+            except TypeError:
+                # datetime safe version
+                return self.left + 0.5 * (self.right - self.left)
+
+
+cdef _interval_like(other):
+    return (hasattr(other, 'left')
+            and hasattr(other, 'right')
+            and hasattr(other, 'closed'))
+
+
+cdef class Interval(IntervalMixin):
+    """
+    Immutable object implementing an Interval, a bounded slice-like interval.
+
+    .. versionadded:: 0.20.0
+
+    Properties
+    ----------
+    left, right : values
+        Left and right bounds for each interval.
+    closed : {'left', 'right', 'both', 'neither'}
+        Whether the interval is closed on the left-side, right-side, both or
+        neither. Defaults to 'right'.
+    """
+
+    cdef readonly object left, right
+    cdef readonly str closed
+
+    def __init__(self, left, right, str closed='right'):
+        # note: it is faster to just do these checks than to use a special
+        # constructor (__cinit__/__new__) to avoid them
+        if closed not in _VALID_CLOSED:
+            raise ValueError("invalid option for 'closed': %s" % closed)
+        if not left <= right:
+            raise ValueError('left side of interval must be <= right side')
+        self.left = left
+        self.right = right
+        self.closed = closed
+
+    def __hash__(self):
+        return hash((self.left, self.right, self.closed))
+
+    def __contains__(self, key):
```

Comment:
@jorisvandenbossche if you see latest, this is changed:

``__contains__`` *only* will do actual ``in`` checking (must be an ``Interval``). I defined ``_is_contained_in`` to be overriden by sub-classes. We could expose this with a ``contains`` method (or simply rename it to be even less confusing).